### PR TITLE
daemon: Remove old health EP state dirs in restore

### PR DESCRIPTION
**Pull request by joestringer**
In the restore logic, when attempting to restore endpoints, if there are
old health endpoint state files left on the filesystem, clean them up as
these endpoints will not be restored, and otherwise they will just hang
around forever.

This is likely to mostly affect non-containerised environments, probably
with the biggest impact being in developer VMs where devs often restart
cilium. Marking for backport because I think it's safe, but it's not pressing.
Happy to drop the backport labels if people feel strongly otherwise.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7249)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

File: daemon/state.go [hunk 1]
```diff
@@ -101,8 +101,18 @@ func (d *Daemon) restoreOldEndpoints(dir string, clean bool) (*endpointRestoreSt
 		// On each restart, the health endpoint is supposed to be recreated.
 		// Hence we need to clean health endpoint state unconditionally.
 		if ep.HasLabels(labels.LabelHealth) {
-			// Completely ignore health endpoint and don't report
-			// it as not restored.
+			// Ignore health endpoint and don't report
+			// it as not restored. But we need to clean up the old
+			// state files, so do this now.
+			healthStateDir := filepath.Join(option.Config.StateDir, ep.StringID())
```
Comment (brb) [inline daemon/state.go]: Maybe we can use `ep.StateDirectoryPath()` here?

**Review by brb — DISMISSED** [review]

**Comment by joestringer** [comment]
test-me-please

EDIT: Hit #7105 yet again. All other tests seem fine.

File: daemon/state.go [hunk 2]
```diff
@@ -101,8 +101,18 @@ func (d *Daemon) restoreOldEndpoints(dir string, clean bool) (*endpointRestoreSt
...
+			// it as not restored. But we need to clean up the old
+			// state files, so do this now.
+			healthStateDir := ep.StateDirectoryPath()
+			scopedLog.WithFields(logrus.Fields{
+				logfields.Path: healthStateDir,
+			}).Debugf("Removing old health endpoint state directory")
+			if err := os.RemoveAll(healthStateDir); err != nil {
+				scopedLog.WithFields(logrus.Fields{
+					logfields.Path: healthStateDir,
+				}).Infof("Cannot clean up old health state directory")
```
Comment (tgraf) [inline daemon/state.go]: Please change this from info to warning

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by tgraf** [comment]
` k8s-1.13.K8sFQDNTest Restart Cilium validate that FQDN is still working` failed

**Comment by joestringer** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Review by brb — APPROVED** [review]

**Comment by joestringer** [comment]
v1.3 used a slightly different approach in `daemon/state.go` which collected a list of endpoints to clean up the state of, then clear them out. That logic is equivalent to the logic proposed in this PR, so v1.3 is not affected. Removed backport tag.

**Pull request merged**